### PR TITLE
Update float16 tutorial

### PR DESCRIPTION
**Pull request by Ishitori**
I have moved float16 tutorial to .MD file format as it seems to be main one in new docs.

I have added a few code examples and fixed existing a little bit, to make sure they can be run by copy & paste.

### CODE REVIEW

File: python/guide/performance/compression/float16.md [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+<!--- specific language governing permissions and limitations -->
+<!--- under the License. -->
+
+# Mixed precision training using float16
+
+In this tutorial we will walk through how one can train deep learning neural networks with mixed precision on supported hardware. We will first see how to use float16 (both with Gluon and Symbolic APIs) and then some techniques on achieving good performance and accuracy.
+
+## Background
+
+The computational resources required for training deep neural networks have been increasing lately, because of growing of the complexity and size of models. Mixed precision training allows us to reduce the utilization of the resources by using lower precision arithmetic. In this approach you can train using 16 bit floating points (half precision) while using 32 bit floating points (single precision) for output buffers of float16 computation. It allows to achieve the same accuracy as training with single precision, while decreasing the required memory and training or inference time. Training using that combination of single and half precision is called mixed precision training.
```
Comment (vishaalkapoor) [inline python/guide/performance/compression/float16.md]: have been lately increasing because of growing complexity and model size.

by using lower precision arithmetic which is computationally less expensive and less costly in terms of space utilization.

Comment (vishaalkapoor) [reply python/guide/performance/compression/float16.md]: - you can train using 16 bit floating point
- while using 32 bit floating point (no s)
- This allows one to achieve the same accuracy ....

Comment (vishaalkapoor) [reply python/guide/performance/compression/float16.md]: "Training using that combination of single and half precision is called mixed precision training." redundant and can be removed.

File: python/guide/performance/compression/float16.md [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
...ta type is a 16 bit floating point representation according to the [IEEE 754 standard](https://ieeexplore.ieee.org/document/4610935). It has a dynamic range where the precision can go from 0.0000000596046 (highest, for values closest to 0) to 32 (lowest, for values in the range 32768-65536). Despite the inherent reduced precision when compared to single precision float (float32), using float16 has many advantages. The most obvious advantages are that you can reduce the size of the model by half allowing the training of larger models and using larger batch sizes. The reduced memory footprint also helps in reducing the pressure on memory bandwidth and lowering communication costs. On hardware with specialized support for float16 computation you can also greatly improve the speed of training and inference. The Volta range of Graphics Processing Units (GPUs) from Nvidia have [Tensor Cores](https://www.nvidia.com/en-us/data-center/tensorcore/) which perform efficient float16 computation. A tensor core allows accumulation of half precision products into single or half precision outputs. For the rest of this tutorial we assume that we are working with Nvidia's Tensor Cores on a Volta GPU.
```
Comment (vishaalkapoor) [inline python/guide/performance/compression/float16.md]: NVidia

File: python/guide/performance/compression/float16.md [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+
+## Prerequisites
+
+- Volta range of Nvidia GPUs (e.g. AWS P3 instance)
```
Comment (vishaalkapoor) [inline python/guide/performance/compression/float16.md]: Link to https://www.nvidia.com/en-us/data-center/volta-gpu-architecture/

File: python/guide/performance/compression/float16.md [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+- Cuda 9 or higher
```
Comment (vishaalkapoor) [inline python/guide/performance/compression/float16.md]: CUDA

File: python/guide/performance/compression/float16.md [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+- CUDNN v7 or higher
```
Comment (vishaalkapoor) [inline python/guide/performance/compression/float16.md]: cuDNN

File: python/guide/performance/compression/float16.md [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+
+This tutorial also assumes understanding of how to train a network with float32. Please refer to other tutorials [here](http://beta.mxnet.io/guide/index.html) to get started with Apache MXNet and Gluon API. This tutorial focuses on the changes needed to switch from float32 to mixed precision and tips on achieving the best performance with mixed precision.
```
Comment (vishaalkapoor) [inline python/guide/performance/compression/float16.md]: train a network with float32 (the default)

Comment (vishaalkapoor) [inline python/guide/performance/compression/float16.md]: Better to include a specific tutorial on training, or crash course or something rather than the guide landing page.

File: python/guide/performance/compression/float16.md [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+
+## Using the Gluon API
+
+### Training or Inference
+
+With Gluon API, you need to take care of three things to convert a model to support float16.
```
Comment (vishaalkapoor) [inline python/guide/performance/compression/float16.md]: to support computation with float16

File: python/guide/performance/compression/float16.md [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+
+1. Cast Gluon Block's parameters and expected input type to float16 by calling the [cast](http://beta.mxnet.io/api/gluon/mxnet.gluon.nn.Block.html#mxnet.gluon.nn.Block.cast) method of the Block representing the network.
```
Comment (vishaalkapoor) [inline python/guide/performance/compression/float16.md]: `Block` here and elsewhere

File: python/guide/performance/compression/float16.md [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+
+```
+net = net.cast('float16')
+```
+
+2. Ensure the data input to the network is of float16 type. If your DataLoader or Iterator produces output in another datatype, then you would have to cast your data. There are different ways you can do this. The easiest would be to use the [astype](http://beta.mxnet.io/api/ndarray/_autogen/mxnet.ndarray.NDArray.astype.html#mxnet.ndarray.NDArray.astype) method of NDArrays.
```
Comment (vishaalkapoor) [inline python/guide/performance/compression/float16.md]: `DataLoader` `Iterator`

File: python/guide/performance/compression/float16.md [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+
+```python
+data = data.astype('float16', copy=False)
+```
+
+If you are using images and DataLoader, you can also use a [Cast transform](http://beta.mxnet.io/api/gluon/_autogen/mxnet.gluon.data.vision.transforms.Cast.html).
+
+3. It is preferable to use **multi_precision mode of optimizer** when training in float16. This mode of optimizer maintains a master copy of weights in float32 even when the training (i.e. forward and backward pass) is in float16. This helps increase precision of the weight updates and can lead to faster convergence for some networks.
```
Comment (vishaalkapoor) [inline python/guide/performance/compression/float16.md]: of the weights

Comment (vishaalkapoor) [reply python/guide/performance/compression/float16.md]: can lead to faster convergence. (Is it for some networks or in some scenarios).

File: python/guide/performance/compression/float16.md [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+
+```python
+optimizer = mx.optimizer.create('sgd', multi_precision=True, lr=0.01)
+```
+
+You can play around with mixed precision using the image classification example [here](https://github.com/apache/incubator-mxnet/blob/master/example/gluon/image_classification.py). We suggest using the Caltech101 dataset option in that example and using a Resnet50_v1 network so you can quickly see the performance improvement and how the accuracy is unaffected. Here's the starter command to run this example.
```
Comment (vishaalkapoor) [inline python/guide/performance/compression/float16.md]: using the image classification [example]

File: python/guide/performance/compression/float16.md [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+
+### Fine-tuning
+
+You can also fine-tune a model, which was originally trained in float32, to use float16. Below is an example of how to fine-tune a pretrained model from the Model Zoo. You would first need to fetch the pretrained network and then cast that network to float16.
+
+```python
+import numpy as np
+import mxnet as mx
+from mxnet.gluon.model_zoo.vision import get_model
+
```
Comment (vishaalkapoor) [inline python/guide/performance/compression/float16.md]: two spaces after import

File: python/guide/performance/compression/float16.md [hunk 13]
```diff
@@ -0,0 +1,243 @@
...
+For a full example, please refer to [resnet.py](https://github.com/apache/incubator-mxnet/blob/master/example/image-classification/symbols/resnet.py) file on GitHub. A small, relevant excerpt from that file is presented below.
+
+```python
+data = mx.sym.Variable(name="data")
+
+if dtype == 'float16':
+    data = mx.sym.Cast(data=data, dtype=np.float16)
+
+// the rest of the network
```
Comment (vishaalkapoor) [inline python/guide/performance/compression/float16.md]: Do you mean # The rest of the network

File: python/guide/performance/compression/float16.md [hunk 14]
```diff
@@ -0,0 +1,243 @@
...
+```
+
+If you don't have the `Caltech256` dataset, you can download it using the script below, and convert it into .rec file format using [im2rec utility file](https://github.com/apache/incubator-mxnet/blob/master/tools/im2rec.py)
+
+```python
+import os
+from os.path import expanduser
+import tarfile
+import mxnet as mx
+
```
Comment (vishaalkapoor) [inline python/guide/performance/compression/float16.md]: two lines

File: python/guide/performance/compression/float16.md [hunk 15]
```diff
@@ -0,0 +1,243 @@
...
+    print('Extracting {} in {}...'.format(archive_file, data_folder))
+    tar = tarfile.open(archive_path)
+    tar.extractall(data_folder)
+    tar.close()
+    print('Data extracted.')
+```
+
+## Example training results
+
+Let us consider training a Resnet50 v1 model on the ImageNet 2012 dataset. For this model, the GPU memory usage is close to the capacity of V100 GPU with a batch size of 128 when using float32. Using float16 allows the use of 256 batch size. Shared below are results using 8 V100 GPUs on a AWS p3.16x large instance.
```
Comment (vishaalkapoor) [inline python/guide/performance/compression/float16.md]: ResNet50V1

Comment (vishaalkapoor) [reply python/guide/performance/compression/float16.md]: an AWS p3.16xlarge (full name) and link to instance info

File: python/guide/performance/compression/float16.md (hunk 11 above)
Comment (vishaalkapoor) [reply python/guide/performance/compression/float16.md]: ResNet50V1 (here and below)
ResNet here and below

File: python/guide/performance/compression/float16.md (hunk 2 above)
Comment (Ishitori) [reply python/guide/performance/compression/float16.md]: As it turns out (I was surprised as well), the proper name of writing is Nvidia with lower cased "v".

File: python/guide/performance/compression/float16.md (hunk 10 above)
Comment (Ishitori) [reply python/guide/performance/compression/float16.md]: I guess scenarios are more generic term, so I put scenarios.

**Comment by Ishitori** [comment]
Thanks for the comments, @vishaalkapoor. I have addressed them all in the main repo - https://github.com/apache/incubator-mxnet/pull/15107 Closing this one as a duplicate.
